### ISSUE TITLE
Memory usage during compilation

### ISSUE THREAD

**Issue by donatello** [snoyberg/file-embed#24]
When I use either this package or wai-app-static to embed a file around 25MB in size (an executable to be served over http), I see that compilation takes over 3-4 GB of RAM and sometimes crashes my laptop. 

Is this expected and can anything be done to make it more efficient? It seems that such high memory usage should not actually be needed during compliation.

**Comment by mgsloan** [comment]
This is probably a GHC issue.  Ideally, Template Haskell should let you directly use a `ByteArray` or similar.  Currently, it is run through a `[Word8]`.  I'm surprised the memory usage would be that high, though.

**Comment by donatello** [comment]
Here is a somewhat minimal demonstration of the issue - https://github.com/donatello/file-embed-exp

Should I report this as an issue with ghc?

**Comment by mgsloan** [comment]
Sure, I think it'd be reasonable.  Unfortunately, base does not have any type like `Bytestring` or `Text` for TH to use.  It'd make sense to have something that could use `ByteArray#` (probably wrapped in a datatype), and something that could use `(Ptr Word8, Int)` or something like that.

**Comment by donatello** [comment]
I've opened an issue here - https://ghc.haskell.org/trac/ghc/ticket/14741

**Comment by steshaw** [comment]
@donatello Would be good to see a solution via GHC but in the meantime, maybe file-embed's [inject system](https://www.stackage.org/haddock/lts-12.5/file-embed-0.0.10.1/Data-FileEmbed.html#g:3) would suit your purpose? The closest thing I can find to a tutorial on it is [here](https://www.yesodweb.com/blog/2011/06/building-better-chm).

**Comment by MaxGabriel** [comment]
@hsyl20 Am I understanding the implications of your work correctly, that every time we use fileEmbed, even on the smallest of files, it costs over a second of compilation time? The github links in your post 404 so I wasn't sure what exactly was being tested.

**Comment by hsyl20** [comment]
The links in my blog post should be fixed soon (i have merged some repos...). The test script was: https://github.com/haskus/packages/blob/ef893138620e8cea1b9ecb2983f5e38b997bd387/haskus-binary/bench.sh
The timings have been obtained with ``time stack exec -- ghc -v0 bench/BenchEmbed.hs -fforce-recomp -O`` and on my setup it took over a second even for the smallest test (but it may not be due to file-embed, my computer is quite slow).

Anyway I have made some patchs that have been integrated into GHC:
* https://gitlab.haskell.org/ghc/ghc/commit/4fa32293c9d2658ce504b8fe6d909db2acf59983
* https://gitlab.haskell.org/ghc/ghc/commit/1d9a1d9fb8fe0a1fea2c44c4246f102ff3e1f3a3

Thanks to them ``file-embed`` should be much faster in the next GHC release (when the native code generator is used, not LLVM).

I also have this merge request that I hope will be merged soon: https://gitlab.haskell.org/ghc/ghc/merge_requests/141
Once it is merged, ``file-embed`` could be patched to use the new ``Bytes`` primitive as can be seen here:
https://gitlab.haskell.org/ghc/ghc/merge_requests/141/diffs#e0300ef07066368947e579d6342b69d6d88f881c_0_18

Then embedding large files will only consists in reading them into a ByteString and passing the ByteString pointer to GHC via TH (``file-embed`` side) and writing them back into another for inclusion into the binary (GHC side): no transformation into ``[Word8]`` anymore \o/.

**Comment by MaxGabriel** [comment]
That’s wonderful! Thank you for those patches @hysl20!

As far as the time to do a single file embed on a small file, thankfully it isn’t 1 second or anything. I removed about 20 from our codebase and saw a 1 second compile time reduction, and that could be just noise.

**Comment by hsyl20** [comment]
The MR has been merged: https://gitlab.haskell.org/ghc/ghc/commit/224a6b864c6aa0d851fcbf79469e5702b1116dbc
Starting with GHC 8.10 (I think) we will be able to use the new approach in TH.

**Comment by saurabhnanda** [comment]
I was pointed to this package in a [reply to a question I asked on Reddit](https://www.reddit.com/r/haskell/comments/gfsd4b/whats_the_packaging_story_for_resuable_webapps_in/fpvkx65/). 

 I'm not too sure if embedding a web-app's static assets (HTML, CSS, JS) directly into the binary is a good idea. I am not aware of any other language/framework doing that, but I may be wrong.

In view of this issue (extremely high memory usage during compilation), would you recommend this approach? (We have 100s of static assets being using in our production app).

**Comment by saurabhnanda** [comment]
Also, this got me thinking, can file embedding be achieved by the use of a pre-processor? Will GHC choke if there is a module with very large strings, eg.

```haskell
module StaticAssets where

index_js :: ByteString
index_js = "...." -- about 800kb of compressed/minified JS

-- and about 50-60 such top-level declarations
```

**Comment by hsyl20** [comment]
`file-embed` should be patched as I proposed in [1] now that TH provides `Bytes` literals [2]. In my tests last year [1] I found that the impact on compilation is then minimal (which the native code generator).

@saurabhnanda using another pre-processor to generate large strings would likely be more costly. With the current approach we bypass the parser and with the approach in [1] we would also bypass TH to Haskell AST string conversion.

[1] https://gitlab.haskell.org/ghc/ghc/issues/14741#note_167616
[2] https://hackage.haskell.org/package/template-haskell-2.16.0.0/docs/Language-Haskell-TH-Lib.html#v:bytesPrimL

**Comment by saurabhnanda** [comment]
@hsyl20 is this the complete patch? If yes, I can try this out in the [odd-jobs job-queue lbrary](https://hackage.haskell.org/package/odd-jobs) and even use this in production and report back.

```
< module Data.FileEmbed
---
> module FileEmbed
49,53c49
< #if MIN_VERSION_template_haskell(2,5,0)
<     , Lit (StringL, StringPrimL, IntegerL)
< #else
<     , Lit (StringL, IntegerL)
< #endif
---
>     , Lit (..)
60a57
> import Language.Haskell.TH
65a63
> import qualified Data.ByteString.Internal as B
154c152,156
< #if MIN_VERSION_template_haskell(2, 8, 0)
---
> #if MIN_VERSION_template_haskell(2, 15, 0)
>       `AppE` LitE (bytesPrimL (
>                 let B.PS ptr off sz = bs
>                 in  mkBytes ptr (fromIntegral off) (fromIntegral sz))))
> #elif MIN_VERSION_template_haskell(2, 8, 0)
```

**However,** we're using GHC 8.4.x in production and the core GHC patch has been applied ONLY to 8.10.x, right? What would it take to back-port the core GHC patch?

**Comment by hsyl20** [comment]
> is this the complete patch? 

Yes. Except that:

```
> #if MIN_VERSION_template_haskell(2, 15, 0)
```
should be:

```
> #if MIN_VERSION_template_haskell(2, 16, 0)
```

> However, we're using GHC 8.4.x in production and the core GHC patch has been applied ONLY to 8.10.x, right? What would it take to back-port the core GHC patch?

It shouldn't be too hard. You have to backport some parts of the following patches:

* TH patch: https://gitlab.haskell.org/ghc/ghc/-/commit/224a6b864c6aa0d851fcbf79469e5702b1116dbc
* NCG .incbin patch: https://gitlab.haskell.org/ghc/ghc/-/commit/1d9a1d9fb8fe0a1fea2c44c4246f102ff3e1f3a3
* Use ByteString to represent Cmm string literals: https://gitlab.haskell.org/ghc/ghc/-/commit/4fa32293c9d2658ce504b8fe6d909db2acf59983

**Comment by snoyberg** [comment]
Is someone interested in submitting a PR for the changes here?

**Comment by hsyl20** [comment]
> Is someone interested in submitting a PR for the changes here?

I'll do it

**Comment by hsyl20** [comment]
Done in #36

**Comment by qrilka** [comment]
Didn't #36 resolve this issue?

**Comment by tschuchortdev** [comment]
@hsyl20 @snoyberg What's the status on this? Can this library be used now with large files without incurring a significant performance hit since #36 has been merged or should a different approach be used?

**Comment by snoyberg** [comment]
I haven't tested this out myself, so I can't vouch one way or the other. But it looks like it should be fixed with #36. I'm going to close the issue (which I should have done over a year ago), if anyone encounters issues please ping again.
